src/core/callsite.c

```diff
@@ -431,3 +432,35 @@ MVMCallsite * MVM_callsite_insert_positional(MVMThreadContext *tc, MVMCallsite *
 
     return new_callsite;
 }
+
+/* Produce a new callsite consisting of the current one with a positional
+ * argument inserted. */
+MVMCallsite * MVM_callsite_replace_positional(MVMThreadContext *tc, MVMCallsite *cs, MVMuint32 idx,
+        MVMCallsiteFlags flag) {
+    /* Can only do this with positional arguments and non-flattening callsite. */
+    if (idx > cs->num_pos)
+        MVM_exception_throw_adhoc(tc, "Cannot replace positional in callsite: index out of range");
+    if (cs->has_flattening)
+        MVM_exception_throw_adhoc(tc, "Cannot transform a callsite with flattening args");
+
+    /* Allocate a new callsite and set it up. */
+    MVMCallsite *new_callsite = MVM_malloc(sizeof(MVMCallsite));
```

Comment:
Could this not have a shortcut for when the new flag is exactly the same as the existing one, so we'd produce an identical callsite anyway?

Reply:
i tried that and got into memory management trouble. we could check if it's interned and in that case we can just return it unchanged, though. otherwise we'll need to clone, or else we'll double-free stuff

Reply:
Ah, fair point.